### PR TITLE
Use kustomize to generate patched CRDs for specific flavors

### PR DESCRIPTION
**Pull request by nkvoll**
Use kustomize to generate patched CRDs for specific flavors

Adds two flavors of the CRDs:

- `default`: Removes validation of `podTemplate`.
- `trivial-versions`: No apiserver-side validation at all. Required for K8s <= 1.13.

This requires the client-side version of the `kubectl` tool to be >= 1.14.

Example:

```shell
$ kubectl kustomize config/crds | wc
   28787  122012 1946531
$ kubectl kustomize config/crds-flavor-default | wc
   11327   49785  739296
$ kubectl kustomize config/crds-flavor-trivial-versions | wc
     149     291    3053
$ kubectl apply -k config/crds-flavor-trivial-versions --dry-run
customresourcedefinition.apiextensions.k8s.io/apmservers.apm.k8s.elastic.co configured (dry run)
customresourcedefinition.apiextensions.k8s.io/elasticsearches.elasticsearch.k8s.elastic.co configured (dry run)
customresourcedefinition.apiextensions.k8s.io/kibanas.kibana.k8s.elastic.co configured (dry run)
```

To install flavored CRDS, set `CRD_FLAVOR` when calling make:

```
$ make install-crds CRD_FLAVOR=trivial-versions
...
customresourcedefinition.apiextensions.k8s.io/apmservers.apm.k8s.elastic.co configured
customresourcedefinition.apiextensions.k8s.io/elasticsearches.elasticsearch.k8s.elastic.co configured
customresourcedefinition.apiextensions.k8s.io/kibanas.kibana.k8s.elastic.co configured
```

If no flavor is specified, `default` will be used.

### Followups

- Upload more than one flavor of the all-in-one yaml from CI. https://github.com/elastic/cloud-on-k8s/issues/1876
- Document different flavors. https://github.com/elastic/cloud-on-k8s/issues/1876

### CODE REVIEW

**Comment by nkvoll** [comment]
@artemnikitin re the CI failure:

```
15:14:57  Error: unknown command "kustomize" for "kubectl"
15:14:57  Run 'kubectl --help' for usage.
15:14:57  unknown command "kustomize" for "kubectl"
```

Would it be possible to upgrade the `kubectl` version on CI? (Should be fine with >= 1.14, with 1.16 being the current latest version)

**Comment by anyasabo** [comment]
Looks like we need to rev `ENV KUBECTL_VERSION 1.13.6` to v1.14+ in a few docker files too for tests to pass, since 1.13 doesn't have kustomize built in.

I think we need to figure out either docs and/or tooling for updating k8s and stack versions, this is definitely not the only time something similar has happened

**Comment by artemnikitin** [comment]
PR with `kubectl` update https://github.com/elastic/cloud-on-k8s/pull/1856

**Comment by anyasabo** [comment]
jenkins test this please

**Comment by artemnikitin** [comment]
I think that PR should be rebased on top of master to get fix...

**Comment by anyasabo** [comment]
@artemnikitin If I'm reading it right it merges it in before running tests, so we should be good there. That said it looks now it is erroring out because it needs `make generate` run against the new max surge description. But it should already be running with kubectl 1.14.


To recap an out of band discussion with @nkvoll :

- I think we want to call the `generate-crds` target inside of `generate`. We can then likely remove the `generate-crds` calls elsewhere. It also may be worth renaming `generate-crds` to `kustomize-crds`. I noticed this because running `make e2e-local CRD_FLAVOR=crds-with-trivial-versions` still failed against a k8s 1.12 cluster as it is only installing the CRDs as they exist on disk (not necessarily post-kustomize).

- We likely want to always generate both a current/preferred CRD, and a backwards compatible CRD (since <1.13 does not support `schema`). We should also update the docs to mention the backwards compatible all-in-one.yaml. Right now in this PR we are just generating the 1.13+ CRD.
  - I think my vote for this is keep the `all-in-one.yaml` as the current version, then create separate files as needed for backwards compatibility. In this case, I'd propose `all-in-one-trivial-versions.yaml`. An alternative of naming it after the k8s version, e.g. `all-in-one-1-12.yaml` seems potentially confusing because it would also be used for 1.11.
  - I think we may also need to update e2e tests against older k8s versions to use these new versions of the CRDs.

I don't think we reached any conclusion, but that's where my thoughts are now at least.

**Comment by nkvoll** [comment]
- [x] Split the PR into two commits so it's more manageable outside of GH "files changed" view
- [x] Validation for `CRD_FLAVOR` in the Makefile
- [x] Validation that all flavors works with kustomize during generation.
- [x] Now generates all in one configs that are flavored. Defaults to uploading the `default` flavor, should likely upload multiple flavors and have documentation to match (noted in the TODO section of the PR description for nw).
- [x] Support for selecting a flavor when running e2e tests. Defaults to `default`.
- [x] It should now be possible to pass `CRD_FLAVOR` to all relevant makefile targets.

File: Makefile [hunk 1]
```diff
@@ -41,8 +41,15 @@ else
...
 
-# Produce CRDs that work back to Kubernetes <1.13 (no version conversion)
-CRD_OPTIONS ?= "crd:trivialVersions=true"
+CRD_OPTIONS ?= "crd"
+
+# CRD_FLAVOR can be used to select specific flavors of CRDs
+CRD_FLAVOR ?= default
+CRD_AVAILABLE_FLAVORS := default trivial-versions
+# verify that the CRD_FLAVOR is valid:
+ifeq ($(filter $(CRD_FLAVOR),$(CRD_AVAILABLE_FLAVORS)),)
```
Comment (anyasabo) [inline Makefile]: nice

**Comment by anyasabo** [comment]
I'm good with this approach 👍 . For the TODOs, do you mean for those to be in follow up PRs? If so, it's worth noting that I think that blocks the next release since we will need them to support <1.13 versions.

I think this is where you might need to change it to get the PR CI tests passing:
https://github.com/elastic/cloud-on-k8s/blob/master/build/ci/pr/Jenkinsfile#L76

Though there might be other places we need to change it outside of the PR one.

**Comment by nkvoll** [comment]
- [x] Jenkinsfiles for e2e tests are updated to use a suitable flavor, depending on the actual version.
- [x] Strip podTemplate for v1alpha1 AND v1beta1 for APM/Kibana
- [x] Generate all all-in-one-flavors by default in the `generate-all-in-one` make target

File: build/ci/e2e/GKE_k8s_versions.jenkinsfile [hunk 2]
```diff
@@ -86,6 +86,10 @@ REGISTRY = eu.gcr.io
 REPOSITORY = "$GCLOUD_PROJECT"
 SKIP_DOCKER_COMMAND = true
 EOF
+        if [[ "${clusterVersion}" < "1.14" ]] then
```
Comment (thbkrkr) [inline build/ci/e2e/GKE_k8s_versions.jenkinsfile]:
```suggestion
        if [[ "${clusterVersion}" < "1.14" ]]; then
```

File: config/crds-flavor-default/apm-and-kibana-podtemplate-patch.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# remove spec.podTemplate
+- op: remove
+  path: /spec/versions/0/schema/openAPIV3Schema/properties/spec/properties/podTemplate
+- op: remove
+  path: /spec/versions/1/schema/openAPIV3Schema/properties/spec/properties/podTemplate
```
Comment (anyasabo) [inline config/crds-flavor-default/apm-and-kibana-podtemplate-patch.yaml]: You can probably set your editor to add the trailing newline for you :)

**Pull request merged**